sequencer/src/state.rs

```diff
@@ -1,47 +1,162 @@
+            proposal.block_merkle_tree_root == block_merkle_tree_root,
+            anyhow::anyhow!(
+                "Invalid Block Root Error: {}, {}",
+                block_merkle_tree_root,
+                proposal.block_merkle_tree_root
+            )
+        );
 
+        // fetch receipts from the l1
+        let receipts = fetch_fee_receipts(parent);
+        for FeeReceipt { recipient, amount } in receipts {
+            // Get the balance in order to add amount, ignoring the proof.
+            match fee_merkle_tree.universal_lookup(recipient) {
+                LookupResult::Ok(balance, _) => fee_merkle_tree
+                    .update(recipient, balance.add(amount))
+                    .unwrap(),
+                // Handle `NotFound` and `NotInMemory` by initializing
+                // state.
+                _ => fee_merkle_tree.update(recipient, amount).unwrap(),
+            };
+        }
+
+        let fee_merkle_tree_root = fee_merkle_tree.commitment();
+        anyhow::ensure!(
+            proposal.fee_merkle_tree_root == fee_merkle_tree_root,
+            anyhow::anyhow!(
+                "Invalid Fee Root Error: {}, {}",
+                fee_merkle_tree_root,
+                proposal.fee_merkle_tree_root
+            )
+        );
+
+        Ok(ValidatedState {
+            block_merkle_tree,
+            fee_merkle_tree,
+        })
+    }
+}
+
+impl HotShotState for ValidatedState {
+    type Error = BlockError;
+    type Instance = NodeState;
     type BlockHeader = Header;
     type BlockPayload = Payload;
 
     type Time = ViewNumber;
 
     fn on_commit(&self) {}
-
+    /// Validate parent against known values (from state) and validate
+    /// proposal descends from parent. Returns updated `ValidatedState`.
     fn validate_and_apply_header(
         &self,
-        _proposed_header: &Self::BlockHeader,
-        _parent_header: &Self::BlockHeader,
-        _view_number: &Self::Time,
+        _instance: &Self::Instance,
+        parent_header: &Self::BlockHeader,
+        proposed_header: &Self::BlockHeader,
     ) -> Result<Self, Self::Error> {
-        todo!("Tracking issue: [https://github.com/EspressoSystems/espresso-sequencer/issues/968]")
+        match self.validate_proposal(parent_header, proposed_header) {
+            // Note that currently only block state is updated.
+            Ok(validated_state) => Ok(validated_state),
+            Err(e) => {
+                tracing::warn!("Invalid Proposal: {}", e);
+                Err(BlockError::InvalidBlockHeader)
+            }
+        }
+    }
+    /// Construct the state with the given block header.
+    ///
+    /// This can also be used to rebuild the state for catchup.
+    fn from_header(block_header: &Self::BlockHeader) -> Self {
+        let fee_merkle_tree = FeeMerkleTree::from_commitment(block_header.fee_merkle_tree_root);
+        let block_merkle_tree = BlockMerkleTree::from_commitment(block_header.fee_merkle_tree_root);
```

Comment:
@tbro isn't this supposed to be `block_merkle_tree_root`?

Reply:
No. from_commitment()` returns a merkle tree that contains only the commitment to be updated later (through catch up I believe).